t/05-scheduler-full.t

```diff
@@ -142,81 +159,187 @@ subtest 'Scheduler worker job allocation' => sub {
+    is @$allocated, 10, "Allocated maximum number of jobs that could have been allocated";
+    is $failures, 2, "Failure count should be to 2, since we took too much time to schedule";
+    is get_scheduler_tick($reactor), 2**$failures * $ENV{OPENQA_SCHEDULER_TIMESLOT}, "Tick is at the expected value";
+    my %jobs;
+    my %w;
+    foreach my $j (@$allocated) {
+        ok !$jobs{$j->{job}}, "Job (" . $j->{job} . ") allocated only once";
+        ok !$w{$j->{worker}}, "Worker (" . $j->{worker} . ") used only once";
+        $w{$j->{worker}}++;
+        $jobs{$j->{job}}++;
+    }
+
+    for my $dup (@duplicated) {
+        for (0 .. 10) {
+            last if $dup->state eq OpenQA::Schema::Result::Jobs::SCHEDULED;
+            sleep 2;
+        }
+        is $dup->state, OpenQA::Schema::Result::Jobs::SCHEDULED, "Job(" . $dup->id . ") back in scheduled state";
+    }
+    dead_workers($schema);
+    kill_service($_, 1) for @workers;
+
+    @workers = ();
+
+    push(@workers, unstable_worker($k->key, $k->secret, "http://localhost:$mojoport", $_, 3)) for (1 .. 50);
+    sleep 5;
+
+    ($allocated, $failures, $no_actions) = scheduler_step($reactor); # Will try to allocate to previous worker and fail!
+    is @$allocated, 0, "All failed allocation on second step - workers were killed";
+    ok $failures >= 8, "Failure count($failures) should be >=8, since we took too much time to schedule";
+    is get_scheduler_tick($reactor), $ENV{OPENQA_SCHEDULER_MAX_BACKOFF}, "Tick is at the expected value";
+
+    for my $dup (@duplicated) {
+        for (0 .. 10) {
+            last if $dup->state eq OpenQA::Schema::Result::Jobs::SCHEDULED;
+            sleep 2;
+        }
+        is $dup->state, OpenQA::Schema::Result::Jobs::SCHEDULED, "Job(" . $dup->id . ") is still in scheduled state";
+    }
+
+    kill_service($_, 1) for @workers;
+};
+
+# This test destroys almost everything.
+subtest 'Simulation of heavy failures' => sub {
     my $allocated;
     my $failures;
+    my $no_actions;
+    my @workers;
+
+    kill_service($wspid);
 
-    my $JOB_SETUP
-      = 'ISO=Core-7.2.iso DISTRI=tinycore ARCH=i386 QEMU=i386 QEMU_NO_KVM=1 '
-      . 'FLAVOR=flavor BUILD=1 MACHINE=coolone QEMU_NO_TABLET=1 INTEGRATION_TESTS=1'
-      . 'QEMU_NO_FDC_SET=1 CDMODEL=ide-cd HDDMODEL=ide-drive VERSION=1 TEST=core PUBLISH_HDD_1=core-hdd.qcow2';
+    dead_workers($schema);
 
-    # schedule job
-    diag client_output($k->key, $k->secret, "http://localhost:$mojoport", "jobs post $JOB_SETUP");
+    trigger_capture_event_loop($reactor);
     scheduler_step($reactor);
-    reset_tick($reactor);
 
-    my $w1_pid = create_worker($k->key, $k->secret, "http://localhost:$mojoport", 4);
-    sleep 1;
-    ($allocated, $failures) = scheduler_step($reactor);
-    is @$allocated, 1;
-    is @{$allocated}[0]->{job},    99983;
-    is @{$allocated}[0]->{worker}, 6;
+    push(@workers, unstable_worker($k->key, $k->secret, "http://localhost:$mojoport", $_, 3)) for (1 .. 10);
+    sleep 5;
```

Comment:
i'll look into if i can avoid the sleep - it would work even without, but makes the following test(s) a bit more accurate and avoids random test fails